### PR TITLE
Substations now explode when destroyed

### PR DESCRIPTION
**Pull request by Peptide90**
Substations now explode and create a destroyed machine which can be deconstructed.

Adds a destructible component and explosive component to substations.
Adds a new entity in machines for a destroyed machine variant.
Adds a construction graph which is hidden purely for the deconstrution of said destroyed machine.
Adds a new texture for a destroyed machine frame which looks slightly burnt out and smashed.

:cl:
- fix: Substations now explode when damaged enough and produce debris. Watch out!

### CODE REVIEW

**Comment by SweptWasTaken** [comment]
If a substation doesn't have any power running through it does it still explode?

**Comment by Peptide90** [comment]
> If a substation doesn't have any power running through it does it still explode?

Very likely. Don't imagine we have a way in yaml for conditions of explosion so that it needs power do we? Conversely, not just power supplied but substations also act as batteries so would need to check that it's discharged (and even then there's an argument that there's still trapped energy / charge in there in the form of springs and gasses).

So yeah, could be a potential "TODO". I know explosions are getting refactored too.

**Comment by PaulRitter** [comment]
imo its fine since these are batteries, hence the explosion coming from the battery, not the powersource

File: Resources/Prototypes/Entities/Structures/Machines/frame.yml [hunk 1]
```diff
@@ -105,3 +105,55 @@
     - type: Appearance
       visuals:
       - type: MachineFrameVisualizer
+
+
+- type: entity
+  id: MachineFrameDestroyed
+  name: destroyed machine frame
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Structures/Machines/frame.yml:112]: IMO should inherit from the BaseDynamicStructure or whatever it's called and be wrenchable (via anchorablecomponent) etc.

File: Resources/Prototypes/Entities/Structures/Machines/frame.yml [hunk 2]
```diff
@@ -105,3 +105,55 @@
...
+  components:
+    - type: Transform
+      anchored: true
+    - type: Physics
+      bodyType: Static
+    - type: Fixtures
+      fixtures:
+      - shape:
+          !type:PhysShapeAabb
+            bounds: "-0.25,-0.5,0.25,0.5"
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Structures/Machines/frame.yml:124]: Definitely an interesting BB...

image

File: Resources/Prototypes/Entities/Structures/Machines/frame.yml [hunk 3]
```diff
@@ -105,3 +105,55 @@
...
+        mass: 25
+        layer:
+          - Impassable
+          - MobImpassable
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Structures/Machines/frame.yml:128]: Delete

File: Resources/Prototypes/Recipes/Construction/Graphs/machines/machine.yml [hunk 4]
```diff
@@ -98,3 +103,18 @@
             - tool: Prying
               doAfter: 0.25
 
+    - node: destroyedmachineframe
```
Comment (metalgearsloth) [inline Resources/Prototypes/Recipes/Construction/Graphs/machines/machine.yml:106]: Should these nodes be camelcase?

**Review by PaulRitter — COMMENTED** [review]
do what sloth said. otherwise lgtm

**Comment by SweptWasTaken** [comment]
Are substations batteries?

**Comment by Pangogie** [comment]
Yes.
image

**Comment by SweptWasTaken** [comment]
> Yes. image

Maybe just me but seems baby-mode to have batteries at every point of contact for power.

**Comment by Pangogie** [comment]
I agree. It's kinda weird for the substation to be a battery.

**Comment by Peptide90** [comment]
> I agree. It's kinda weird for the substation to be a battery.

Someone was in talk with PJB recently about rebalancing them. No idea why they currently act as a buffer like that for downstream but I agree if they do have a battery it should be seconds of backup buffer rather than storage like they are now. More importance to SMES. But not relevant to this PR.

Will do the changes requested when I get time over the next week before I work on my new PRs.

**Review by PaulRitter — APPROVED** [review]

File: Resources/Prototypes/Recipes/Construction/Graphs/machines/machine.yml [hunk 5]
```diff
@@ -98,3 +103,18 @@
...
+          steps:
+            - tool: Welding
+              doAfter: 5            
+          completed:
+            - !type:SpawnPrototype
+              prototype: SheetSteel1
+              amount: 5
+            - !type:SpawnPrototype
+              prototype: CableHVStack1
+              amount: 2
```
Comment (ElectroJr) [inline Resources/Prototypes/Recipes/Construction/Graphs/machines/machine.yml:119]: I'm not sure why a destroyed frame is constructible in the first place, but currently there's an issue where it takes only 5 steel sheets to create, but leaves behind 5 sheets & 2 HV wires when deconstructed. It should either not be constructible, not leave behind HV wires, or both.

Comment (Peptide90) [reply Resources/Prototypes/Recipes/Construction/Graphs/machines/machine.yml:119]: > 

Currently the only thing using the destoyed machine frame is the substation and I wanted it to drop some HV cables (whereas everything else is destroyed in the explosion). I can just remove that and say only the steel is salvegable.

The frame itself is not actually constructible but in order for deconstruction you need a complete graph from start to end. Same as my AME deconstruction PR.

**Pull request merged**